crates/bootstrap_srv/src/test.rs

```diff
@@ -0,0 +1,748 @@
+        created_at,
+        expires_at,
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidExpiresAt"));
+}
+
+#[test]
+fn reject_expired_at_too_long() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let created_at = crate::now();
+    let expires_at =
+        created_at + std::time::Duration::from_secs(60 * 40).as_micros() as i64;
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        created_at,
+        expires_at,
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidExpiresAt"));
+}
+
+#[test]
+fn reject_bad_sig() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        signature: Some("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidSignature"));
+}
+
+#[test]
+fn reject_bad_agent_pub_key() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        // only 31 characters... and obviously the wrong key : )
+        final_agent_pk: Some("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidAgentPubKey"));
+}
+
+#[test]
+fn default_storage_rollover() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let addr = s.listen_addr();
+    let mut test_prop: u32 = 0;
+    let mut put_info = move || {
+        use base64::prelude::*;
+        let mut agent_seed = [0; 32];
+        agent_seed[..4].copy_from_slice(&test_prop.to_le_bytes());
+        let agent_seed = BASE64_URL_SAFE_NO_PAD.encode(agent_seed);
+        PutInfo {
+            addr,
+            agent_seed: &agent_seed,
+            test_prop: &format!("{test_prop}"),
+            ..Default::default()
+        }
+        .call()
+        .unwrap();
+        test_prop += 1;
+    };
+
+    let addr = s.listen_addr();
+    let get = move || {
+        let addr = format!("http://{:?}/bootstrap/{}", addr, S1);
+        let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+        let res: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+        res.into_iter().map(|m| m.test_prop).collect::<Vec<_>>()
+    };
+
+    for _ in 0..32 {
+        put_info();
+    }
+
+    let res = get();
+
+    assert_eq!(
+        res.as_slice(),
+        &[
+            "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12",
+            "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23",
+            "24", "25", "26", "27", "28", "29", "30", "31",
+        ]
+    );
+
+    for _ in 0..32 {
+        put_info();
+    }
+
+    let res = get();
+
+    assert_eq!(
```

Comment:
```suggestion
   // Agents put beyond MAX_STORAGE will be insert at MAX_STORAGE  / 2, replacing the existing agent at that place. That is here 32 / 2 = 16, so from 17 onward agents > 32 are replaced. With 32 agents this happens twice, first from 33-47 and then from 48-63.
    assert_eq!(
```

Reply:
```suggestion
    // Agents put beyond MAX_STORAGE will be insert at MAX_STORAGE  / 2,
    // replacing the existing agent at that place. That is here 32 / 2 = 16,
    // so from 17 onward agents > 32 are replaced. With 32 agents this happens
    // twice, first from 33-47 and then from 48-63.
    assert_eq!(
```

:+1: - but reformat to 80 character width